tests/samples: drivers: adc: rework tests filtering and configuration

JarmouniA opened the pull request:
Make tests/samples test scenarios depend on the existence of required DT elements (aliases, "zephyr,user" properties...).

Motivation behind the filtering changes:
The number of board overlays in ADC tests/samples is completely out of control in my opinion. And the reason is: most of their test scenarios get triggered, & fail every single time a board PR is opened and has the `adc` tag in `[board].yaml` (& almost always it does because most people assume those tags are a general list of supported features..., not test dedicated), then an overlay is added, as a fix for CI, eventhough the exact ADC IP/driver is already tested elsewhere on many boards.

JarmouniA committed:
scripts: pylib: twister: expr_parser: add dt_node_has_prop expression

Add 'dt_node_has_prop' expression to use the existence of a node property
as a filter in Twister test configuration yaml files.
First argument must be a node alias (in aliases node) or a node path,
not a nodelabel.

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

hakehuang commented on samples/drivers/adc/adc_dt/sample.yaml:
```diff
@@ -1,51 +1,26 @@
+  tags:
+    - adc
+  depends_on: adc
+  filter: dt_node_has_prop("/zephyr,user","io-channels")
```
> Can you clarify what you mean by clean?

the adc should not be enabled by default, it shall be enabled by application with CONFIG_ADC, or with dts with adc node status.

hakehuang replied:
> That's not a problem in my opinion, we do not aim to build the sample for every single platform in CI. A platform for each ADC IP/driver is sufficient.

what if a platfrom not in `integration_paltform` fails in sample build, is it a valid bug, or we do not support?

hakehuang replied:
when built by a user?

butok commented:
> > > Remove the overlays of redundant platforms that are testing the same IP/Driver, and only keep 1 platform from each SoC series.
> > 
> > 
> > @JarmouniA Please clarify. Does it mean that building of the samples/tests for the "redundant" platforms will be failed?
> 
> If you mean in CI, they won't be built anymore because of the filtering in test conf yaml file.

Not just CI.
Since this PR removes many platform overlays, the "west build" for removed platforms is not possible for these samples.

JarmouniA committed:
tests: drivers: adc: enhance test scenarios filtering

Make ADC tests' test scenarios depend on the existence of
required DT elements.

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

JarmouniA committed:
samples: drivers: adc: enhance test scenarios filtering

Make ADC samples' test scenarios depend on the existence of
required DT elements.

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

anangl commented on scripts/pylib/twister/expr_parser.py:
```diff
@@ -280,6 +280,17 @@ def ast_expr(ast, env, edt):
             return True
         return False
+    elif ast[0] == "dt_node_has_prop":
+        # 1st arg 'label' must be a valid node alias or a node path
+        label = ast[1][0]
+        try:
+            node = edt.get_node(label)
```
Why not name this variable `path` (as the `get_node()` parameter is named)? `label` is a bit confusing.
